src/ResolvedApi.ts

```diff
@@ -0,0 +1,236 @@
+   */
+  form(formId: string): SearchForm | null {
+    const form = this.data.forms[formId];
+    if (form) {
+      return new SearchForm(form, this.httpClient);
+    }
+    return null;
+  }
+
+  everything(): SearchForm {
+    const f = this.form('everything');
+    if (!f) throw new Error('Missing everything form');
+    return f;
+  }
+
+  /**
+   * The ID of the master ref on this prismic.io API.
+   * Do not use like this: searchForm.ref(api.master()).
+   * Instead, set your ref once in a variable, and call it when you need it; this will allow to change the ref you're viewing easily for your entire page.
+   */
+  master(): string {
+    return this.masterRef.ref;
+  }
+
+  /**
+   * Returns the ref ID for a given ref's label.
+   * Do not use like this: searchForm.ref(api.ref("Future release label")).
+   * Instead, set your ref once in a variable, and call it when you need it; this will allow to change the ref you're viewing easily for your entire page.
+   */
+  ref(label: string): string | null {
+    const ref = this.data.refs.filter(ref => ref.label === label)[0];
+    return ref ? ref.ref : null;
+  }
+
+  currentExperiment(): Experiment | undefined {
+    return this.experiments.current();
+  }
+
+  /**
+   * Query the repository
+   */
+  query(q: string | string[], optionsOrCallback: QueryOptions | RequestCallback<ApiSearchResponse>, cb: RequestCallback<ApiSearchResponse> = () => {}): Promise<ApiSearchResponse> {
+    const { options, callback } = typeof optionsOrCallback === 'function'
+        ? { options: {} as QueryOptions, callback: optionsOrCallback }
+    : { options: optionsOrCallback || {}, callback: cb };
+
+    let form = this.everything();
+    for (const key in options) {
+      form = form.set(key, options[key]);
+    }
+    if (!options.ref) {
+      // Look in cookies if we have a ref (preview or experiment)
+      let cookieString = '';
+      if (this.options.req) { // NodeJS
+        cookieString = this.options.req.headers['cookie'] || '';
+      } else if (typeof window !== 'undefined' && window.document) { // Browser
+        cookieString = window.document.cookie || '';
+      }
+      const cookies = Cookies.parse(cookieString);
+      const previewRef = cookies[PREVIEW_COOKIE];
+      const experimentRef = this.experiments.refFromCookie(cookies[EXPERIMENT_COOKIE]);
+      form = form.ref(previewRef || experimentRef || this.masterRef.ref);
+    }
+    if (q) {
+      form.query(q);
+    }
+    return form.submit(callback);
+  }
+
+  /**
+   * Retrieve the document returned by the given query
+   * @param {string|array|Predicate} the query
+   * @param {object} additional parameters. In NodeJS, pass the request as 'req'.
+   * @param {function} callback(err, doc)
+   */
+  queryFirst(q: string | string[], optionsOrCallback: QueryOptions | RequestCallback<Document>, cb?: RequestCallback<Document>): Promise<Document> {
+    const { options, callback } = typeof optionsOrCallback === 'function'
```

Comment:
Would be great to prevent doing this piece of code everywhere at some point.

Reply:
Does it means removing callback support?

Reply:
maybe have a helper to do this or something less painful or yes maybe simply drop callback support I don't know yet what's the best